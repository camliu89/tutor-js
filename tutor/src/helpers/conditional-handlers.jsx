import React from 'react';
import invariant from 'invariant';
import { extend, last, first } from 'lodash';
import { Redirect } from 'react-router-dom';
import { asyncComponent } from './async-component';
import { CourseNotFoundWarning } from '../components/course-not-found-warning';
import Courses from '../models/courses-map';
import { OXMatchByRouter } from 'shared';

const StudentDashboard = asyncComponent(
  () => import('../screens/student-dashboard')
);

const getConditionalHandlers = (Router) => {
  const MatchForTutor = OXMatchByRouter(Router, null, 'TutorRouterMatch');

  const renderDashboard = (props) => {
    const { courseId } = props.params;

    extend(props, { courseId });
    const course = Courses.get(courseId);

    if (!course) {
      return <CourseNotFoundWarning />;
    }

    if (!props.match.isExact) {
      return (
        <MatchForTutor {...props} />
      );
    }

    if (course.isTeacher) {
      if (course.is_concept_coach) {
        return (
          <CCDashboard courseId={courseId} {...props} />
        );
      } else {
        return (
          <Redirect
            to={{
              pathname: Router.makePathname('viewTeacherDashboard', props.params),
              query: Router.currentQuery(),
            }} />
        );
      }
    } else {
      if (course.is_concept_coach) {
        return (
          <CCStudentRedirect courseId={courseId} />
        );
      } else {
        return <StudentDashboard {...props} />;
      }
    }
  };
  // eslint-disable-next-line react/prop-types
  const legacyReferenceBookRedirect = ({ params }) => {
    const parts = params.parts.split('/');
    const course = Courses.get(first(parts));
    invariant(course, `Did not find course for params '${params}'`);
    const chapterSection = parts.length > 1 ? last(parts) : 0;
    return (
      <Redirect
        to={{
          pathname: Router.makePathname(
            chapterSection ? 'viewReferenceBookSection' : 'viewReferenceBook',
            {
              ecosystemId: course.ecosystem_id,
              chapterSection,
            }),
          query: Router.currentQuery(),
        }}
      />
    );
  };

  return {
    dashboard() { return renderDashboard; },
    legacyReferenceBookRedirect() { return legacyReferenceBookRedirect; },
  };
};

export { getConditionalHandlers };
