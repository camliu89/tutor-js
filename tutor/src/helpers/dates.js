import moment from 'moment';
import 'twix';
import 'moment-timezone';
import { compact, reduce, min, max, map } from 'lodash';
import TimeHelper from './time';

export function getDay(oneMoment) {
  return moment(oneMoment)
    .startOf('day')
    .twix(moment(oneMoment).endOf('day'), { allDay: true });
}

export function getDurationFromMoments(listOfMoments) {
  return reduce(
    listOfMoments,
    (current, next) => current.union(getDay(next)),
    getDay(listOfMoments[0])
  );
}

export function findEarliest(dateThings) {
  return min(compact(map(dateThings, d =>
    moment.isMoment(d) ? d ? d.toDate() : new Date(Date.parse(d)) : null,
  )));
}

export function findLatest(dateThings) {
  return max(compact(map(dateThings, d =>
    moment.isMoment(d) ? d ? d.toDate() : new Date(Date.parse(d)) : null,
  )));
}


export function dateWithUnchangedTime(date, t) {
  if (moment.isMoment(date)) {
    date = date.format(TimeHelper.ISO_DATE_FORMAT);
  }
  return TimeHelper.getDateOnly(date);
}
