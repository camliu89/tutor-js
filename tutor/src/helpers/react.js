import React from 'react';
import ReactDOM from 'react-dom';
import cn from 'classnames';
import { propTypes as mobxPropTypes } from 'mobx-react';
import PropTypes from 'prop-types';
import invariant from 'invariant';
import styled, { css } from 'styled-components';
import { idType } from 'shared';

export {
  React, cn, ReactDOM, mobxPropTypes, PropTypes,
  invariant, styled, css, idType,
};
export { observable, action, computed } from 'mobx';
export { observer, inject, Provider } from 'mobx-react';
