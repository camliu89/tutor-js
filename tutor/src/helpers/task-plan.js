import { isEmpty, get, pick, first, every, reduce, map } from 'lodash';
import moment from 'moment';
import Time from '../models/time';

export default {

  // returns the opens_at and due_at dates combined if they're identical,
  // separated by target_id if not
  dates(plan, options) {
    if (options == null) { options = {}; }
    if (isEmpty( get(plan, 'tasking_plans' ))) { return {}; }
    const pickFn = options.only ? tp => pick(tp, options.only) : tp => pick(tp, 'opens_at', 'due_at');

    const firstTasking = first( plan.tasking_plans );
    const isIdentical = every(plan.tasking_plans, function(tp) {
      if (options.only != null) {
        return tp[options.only] === firstTasking[options.only];
      } else {
        return (tp.opens_at === firstTasking.opens_at) && (tp.due_at === firstTasking.due_at);
      }
    });
    if (isIdentical) {
      return { all: pickFn(firstTasking) };
    } else {
      return reduce( plan.tasking_plans, function(memo, tp) {
        memo[tp.target_id] = pickFn(tp);
        return memo;
      }
      , {});
    }
  },

  calendarParams(course) {
    let date;
    const dueAtRange = course
      .teacherTaskPlans.array
      .map(plan => plan.dateRanges.due.start)
      .sort();

    if (dueAtRange.length) {
      date = dueAtRange[0];
    } else {
      date = Time.now;
    }
    return {
      to: 'calendarByDate',
      params: {
        courseId: course.id,
        date: moment(date).format('YYYY-MM-DD'),
      },
    };
  },

  earliestDueDate(plan) {
    const dates = map(get(plan, 'tasking_plans'), 'due_at');
    return first(dates.sort()) || '';
  },

};
