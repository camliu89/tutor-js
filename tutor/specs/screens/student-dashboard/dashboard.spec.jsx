import { React, Router, TimeMock } from '../../helpers';
import Dashboard from '../../../src/screens/student-dashboard/dashboard';
import Factory from '../../factories';
import { bootstrapCoursesList } from '../../courses-test-data';

describe('Student Dashboard', () => {
  let props;

  TimeMock.setTo('2015-10-14T12:00:00.000Z');

  beforeEach(() => {
    const course = Factory.course();
    bootstrapCoursesList();
    Factory.studentTasks({ course, attributes: { now: new Date('2015-10-21T12:00:00.000Z') } });
    course.studentTasks.fetch = jest.fn();
    props = {
      course,
      params: {},
    };
  });

  it('matches snapshot', function() {
    props.course.studentTasks.all_tasks_are_ready = false;
    props.course.primaryRole.joined_at = new Date('2015-09-14T12:00:00.000Z');
    expect.snapshot(<Router><Dashboard {...props} /></Router>).toMatchSnapshot();
  });

  it('displays as loading', () => {
    props.course.studentTasks.all_tasks_are_ready = false;
    props.course.primaryRole.joined_at = new Date('2015-10-14T12:00:00.000Z');
    const dash = mount(<Router><Dashboard {...props} /></Router>);
    expect(dash).toHaveRendered('ThisWeekCard Card[className="empty pending"]');
    expect.snapshot(<Router><Dashboard {...props} /></Router>).toMatchSnapshot();
  });

  it('fetches on mount', () => {
    const dash = mount(<Router><Dashboard {...props} /></Router>);
    expect(props.course.studentTasks.fetch).toHaveBeenCalled();
    dash.unmount();
  });

});
