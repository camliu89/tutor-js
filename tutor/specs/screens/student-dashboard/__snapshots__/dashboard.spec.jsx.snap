// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Student Dashboard displays as loading 1`] = `
.c0 {
  margin-right: 0.5rem;
  margin-left: 0.5rem;
}

<div
  className="student-dashboard"
>
  <div
    className="openstax-notifications-bar"
  />
  <div
    className="course-title-banner"
    data-appearance="testing"
    data-book-title="Testing"
    data-is-preview={false}
    data-term="Spring 2017"
    data-title="Local Test Course Two"
  >
    <div
      className="book-title"
    >
      <h1
        className="book-title-text"
      >
        Local Test Course Two
      </h1>
    </div>
    <div
      className="course-term"
    >
      Spring 2017
    </div>
  </div>
  <div
    className="container"
    data-tour-region-id="student-dashboard"
  >
    <div
      className="row"
    >
      <div
        className="col-lg-9 col-md-8 col-12"
      >
        <nav
          className="tutor-tabs"
        >
          <ul
            className="nav nav-tabs"
            role="tablist"
          >
            <li
              aria-selected={true}
              className="active"
              role="tab"
            >
              <a
                href="#"
                onClick={[Function]}
              >
                <h2>
                  This Week
                </h2>
              </a>
            </li>
            <li
              aria-selected={false}
              className=""
              role="tab"
            >
              <a
                href="#"
                onClick={[Function]}
              >
                <h2>
                  All Past Work
                </h2>
              </a>
            </li>
          </ul>
        </nav>
        <div
          className="tab-pane active"
          role="tabpanel"
        >
          <div
            className="empty pending card"
          >
            <svg
              aria-hidden="true"
              className="svg-inline--fa fa-spinner fa-w-16 fa-spin ox-icon spinner c0"
              data-icon="spinner"
              data-prefix="fas"
              role="img"
              style={Object {}}
              viewBox="0 0 512 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M304 48c0 26.51-21.49 48-48 48s-48-21.49-48-48 21.49-48 48-48 48 21.49 48 48zm-48 368c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48-21.49-48-48-48zm208-208c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48-21.49-48-48-48zM96 256c0-26.51-21.49-48-48-48S0 229.49 0 256s21.49 48 48 48 48-21.49 48-48zm12.922 99.078c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48c0-26.509-21.491-48-48-48zm294.156 0c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48c0-26.509-21.49-48-48-48zM108.922 60.922c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48-21.491-48-48-48z"
                fill="currentColor"
                style={Object {}}
              />
            </svg>
             Preparing assignments for your course.  This can take up to 10 minutes.
          </div>
          <div
            className="empty pending upcoming card"
          >
            <svg
              aria-hidden="true"
              className="svg-inline--fa fa-spinner fa-w-16 fa-spin ox-icon spinner c0"
              data-icon="spinner"
              data-prefix="fas"
              role="img"
              style={Object {}}
              viewBox="0 0 512 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M304 48c0 26.51-21.49 48-48 48s-48-21.49-48-48 21.49-48 48-48 48 21.49 48 48zm-48 368c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48-21.49-48-48-48zm208-208c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48-21.49-48-48-48zM96 256c0-26.51-21.49-48-48-48S0 229.49 0 256s21.49 48 48 48 48-21.49 48-48zm12.922 99.078c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48c0-26.509-21.491-48-48-48zm294.156 0c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48c0-26.509-21.49-48-48-48zM108.922 60.922c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48-21.491-48-48-48z"
                fill="currentColor"
                style={Object {}}
              />
            </svg>
             Preparing assignments for your course.  This can take up to 10 minutes.
          </div>
        </div>
      </div>
      <div
        className="sidebar col-lg-3 col-md-4 col-12"
      >
        <div
          className="actions-box loadable is-loading"
        >
          Loading progress information...
        </div>
        <div
          className="actions-box browse-the-book card"
        >
          <a
            aria-label="Browse course reference book"
            className="browse-the-book"
            data-appearance="college_physics"
            href="/book/1"
            onClick={[Function]}
            target="_blank"
          >
            <div>
              Browse the Book
            </div>
          </a>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Student Dashboard matches snapshot 1`] = `
.c0 {
  margin-right: 0.5rem;
  margin-left: 0.5rem;
}

<div
  className="student-dashboard"
>
  <div
    className="openstax-notifications-bar viewable"
  >
    <div
      className="notification missing-student-id"
    >
      <span
        className="body"
      >
        <svg
          aria-hidden="true"
          className="svg-inline--fa fa-info-circle fa-w-16 ox-icon info-circle c0"
          data-icon="info-circle"
          data-prefix="fas"
          role="img"
          style={Object {}}
          viewBox="0 0 512 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M256 8C119.043 8 8 119.083 8 256c0 136.997 111.043 248 248 248s248-111.003 248-248C504 119.083 392.957 8 256 8zm0 110c23.196 0 42 18.804 42 42s-18.804 42-42 42-42-18.804-42-42 18.804-42 42-42zm56 254c0 6.627-5.373 12-12 12h-88c-6.627 0-12-5.373-12-12v-24c0-6.627 5.373-12 12-12h12v-64h-12c-6.627 0-12-5.373-12-12v-24c0-6.627 5.373-12 12-12h64c6.627 0 12 5.373 12 12v100h12c6.627 0 12 5.373 12 12v24z"
            fill="currentColor"
            style={Object {}}
          />
        </svg>
        <span>
          To get credit for your work, add your student ID.
        </span>
        <a
          className="action"
          onClick={[Function]}
        >
          Add Student ID
        </a>
      </span>
    </div>
  </div>
  <div
    className="course-title-banner"
    data-appearance="testing"
    data-book-title="Testing"
    data-is-preview={false}
    data-term="Spring 2017"
    data-title="Local Test Course One"
  >
    <div
      className="book-title"
    >
      <h1
        className="book-title-text"
      >
        Local Test Course One
      </h1>
    </div>
    <div
      className="course-term"
    >
      Spring 2017
    </div>
  </div>
  <div
    className="container"
    data-tour-region-id="student-dashboard"
  >
    <div
      className="row"
    >
      <div
        className="col-lg-9 col-md-8 col-12"
      >
        <nav
          className="tutor-tabs"
        >
          <ul
            className="nav nav-tabs"
            role="tablist"
          >
            <li
              aria-selected={true}
              className="active"
              role="tab"
            >
              <a
                href="#"
                onClick={[Function]}
              >
                <h2>
                  This Week
                </h2>
              </a>
            </li>
            <li
              aria-selected={false}
              className=""
              role="tab"
            >
              <a
                href="#"
                onClick={[Function]}
              >
                <h2>
                  All Past Work
                </h2>
              </a>
            </li>
          </ul>
        </nav>
        <div
          className="tab-pane active"
          role="tabpanel"
        >
          <div
            className="-this-week card"
          >
            <div
              className="row labels"
            >
              <div
                className="col-sm-7 col-12"
              >
                <span
                  className="date-range"
                >
                  <time>
                    Oct 12, 2015
                  </time>
                  <span>
                    â€“
                  </span>
                  <time>
                    Oct 18, 2015
                  </time>
                </span>
              </div>
              <div
                className="progress-label col-sm-3 col-5 offset-sm-0 offset-2"
              >
                Progress
              </div>
              <div
                className="due-at-label col-sm-2 col-5"
              >
                Due
              </div>
            </div>
            <a
              aria-label="Work on reading: dot-com recontextualize communities"
              className="task row reading"
              data-event-id={1}
              href="/course/1/task/1"
              onClick={[Function]}
              tabIndex={-1}
            >
              <div
                className="column-icon col-sm-1 col-2"
              >
                <i
                  aria-label="reading icon"
                  className="icon icon-lg icon-reading"
                />
              </div>
              <div
                className="title col-sm-6 col-10"
              >
                dot-com recontextualize communities
              </div>
              <div
                className="feedback col-sm-3 col-5"
              >
                <span>
                  Not started
                </span>
              </div>
              <div
                className="due-at col-sm-2 col-5"
              >
                <time>
                  Oct 18, 7:00am
                </time>
              </div>
            </a>
            <a
              aria-label="Work on external: out-of-the-box embrace supply-chains"
              className="task row external"
              data-event-id={2}
              href="/course/1/task/2"
              onClick={[Function]}
              tabIndex={-1}
            >
              <div
                className="column-icon col-sm-1 col-2"
              >
                <i
                  aria-label="external icon"
                  className="icon icon-lg icon-external"
                />
              </div>
              <div
                className="title col-sm-6 col-10"
              >
                out-of-the-box embrace supply-chains
              </div>
              <div
                className="feedback col-sm-3 col-5"
              >
                <span>
                  Not started
                </span>
              </div>
              <div
                className="due-at col-sm-2 col-5"
              >
                <time>
                  Oct 18, 7:00am
                </time>
              </div>
            </a>
            <a
              aria-label="Work on event: B2B deliver content"
              className="task row event"
              data-event-id={3}
              href="/course/1/task/3"
              onClick={[Function]}
              tabIndex={-1}
            >
              <div
                className="column-icon col-sm-1 col-2"
              >
                <i
                  aria-label="event icon"
                  className="icon icon-lg icon-event"
                />
              </div>
              <div
                className="title col-sm-6 col-10"
              >
                B2B deliver content
              </div>
              <div
                className="feedback col-sm-3 col-5"
              >
                <span>
                  
                </span>
              </div>
              <div
                className="due-at col-sm-2 col-5"
              >
                <time>
                  Oct 18, 7:00am
                </time>
              </div>
            </a>
            <a
              aria-label="Work on homework: virtual strategize infomediaries"
              className="task row homework"
              data-event-id={4}
              href="/course/1/task/4"
              onClick={[Function]}
              tabIndex={-1}
            >
              <div
                className="column-icon col-sm-1 col-2"
              >
                <i
                  aria-label="homework icon"
                  className="icon icon-lg icon-homework"
                />
              </div>
              <div
                className="title col-sm-6 col-10"
              >
                virtual strategize infomediaries
              </div>
              <div
                className="feedback col-sm-3 col-5"
              >
                <span>
                  <span
                    className="tour-anchor"
                    data-tour-anchor-id="assignment-progress-status"
                  >
                    10/11 answered
                  </span>
                </span>
              </div>
              <div
                className="due-at col-sm-2 col-5"
              >
                <time>
                  Oct 18, 7:00am
                </time>
              </div>
            </a>
          </div>
          <div
            className="empty upcoming card"
          >
            No upcoming assignments
          </div>
        </div>
      </div>
      <div
        className="sidebar col-lg-3 col-md-4 col-12"
      >
        <div
          className="actions-box loadable is-loading"
        >
          Loading progress information...
        </div>
        <div
          className="actions-box browse-the-book card"
        >
          <a
            aria-label="Browse course reference book"
            className="browse-the-book"
            data-appearance="college_physics"
            href="/book/1"
            onClick={[Function]}
            target="_blank"
          >
            <div>
              Browse the Book
            </div>
          </a>
        </div>
      </div>
    </div>
  </div>
</div>
`;
